crates/transpiler/src/passes/commutative_optimization.rs

```
@@ -0,0 +1,506 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::f64::consts::PI;
+
+use num_complex::Complex64;
+use pyo3::prelude::*;
+use pyo3::{Bound, PyResult, pyfunction, wrap_pyfunction};
+use smallvec::smallvec;
+
+use crate::commutation_checker::{CommutationChecker, get_matrix};
+use crate::passes::remove_identity_equiv::{can_remove, is_identity_equiv};
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{
+    Operation, OperationRef, Param, StandardGate, multiply_param, radd_param,
+};
+
+use qiskit_circuit::VarsMode;
+use qiskit_circuit::packed_instruction::PackedInstruction;
+
+/// Holds the action for each node in the original DAGCircuit.
+#[derive(Clone, Debug)]
+enum NodeAction {
+    /// The node's instruction is unchanged, and can be retrieved from the circuit.
+    Keep,
+    /// The node's instruction can be replaced by this representative
+    /// (the second parameter specifies the global phase update).
+    /// However, unless this representative gate is removed or merged,
+    /// we will add the original instruction to the output circuit.
+    Canonical(PackedInstruction, Param),
+    /// The node's instruction has been removed.
+    Drop,
+    /// The node's instruction has been replaced by the current instruction.
+    Replace(PackedInstruction),
+}
+
+/// Returns true if the two parameter lists are equal.
+fn compare_params(params1: &[Param], params2: &[Param]) -> bool {
+    if params1.len() != params2.len() {
+        return false;
+    }
+
+    for (p1, p2) in params1.iter().zip(params2.iter()) {
+        match p1.eq(p2) {
+            Ok(true) => continue,
+            Ok(false) => return false,
+            Err(_) => return false,
```

If an error occurs this likely shows something with the Python objects went wrong and might indicate a deeper failure -- should we just propagate the error?

---

Done.